src/migrations/20240119000000-exclude_deleted_from_indexes.js

```diff
@@ -0,0 +1,52 @@
+const {
+  prepMigration,
+} = require('../lib/migration');
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+  async up(queryInterface, _Sequelize) {
+    await queryInterface.sequelize.transaction(async (transaction) => {
+      const sessionSig = __filename;
+      await prepMigration(queryInterface, transaction, sessionSig);
+      await queryInterface.sequelize.query(
+        `
+        -- In tables that use Sequelize's built-in "deletedAt" column, exclude deleted records
+        -- from unique indexes that don't include the pkey
+        
+        ALTER TABLE "ActivityReportApprovers" DROP CONSTRAINT "unique_activityReportId_userId";
+        CREATE UNIQUE INDEX "unique_activityReportId_userId"
+        ON "ActivityReportApprovers" ("activityReportId", "userId")
+        WHERE "deletedAt" IS NULL;
+
+        ALTER TABLE "CollaboratorTypes" DROP CONSTRAINT "CollaboratorTypes_name_validForId_unique";
+        CREATE UNIQUE INDEX "CollaboratorTypes_name_validForId_unique"
+        ON "CollaboratorTypes" (name, "validForId")
+        WHERE "deletedAt" IS NULL;
+
+        ALTER TABLE "Courses" DROP CONSTRAINT "Courses_name_unique";
+        CREATE UNIQUE INDEX "Courses_name_unique"
+        ON "Courses" (name)
+        WHERE "deletedAt" IS NULL;
+
+        ALTER TABLE "GoalCollaborators" DROP CONSTRAINT "GoalCollaborators_goalId_userId_collaboratorTypeId_unique";
+        CREATE UNIQUE INDEX "GoalCollaborators_goalId_userId_collaboratorTypeId_unique"
+        ON "GoalCollaborators" ("goalId", "userId", "collaboratorTypeId")
+        WHERE "deletedAt" IS NULL;
+
+        ALTER TABLE "ObjectiveCollaborators" DROP CONSTRAINT "ObjectiveCollaborators_objectiveId_userId_collaboratorTypeId_un";
+        CREATE UNIQUE INDEX "ObjectiveCollaborators_objectiveId_userId_collaboratorTypeId_un"
+        ON "ObjectiveCollaborators" ("objectiveId", "userId", "collaboratorTypeId")
+        WHERE "deletedAt" IS NULL;
+
```

Comment:
Topics seems to be missing

Reply:
Maybe Topic's unique constraint is better as it uses deletedAt as part of the unique index instead of a where clause. That allows it to still be used as a constraint.
`CREATE UNIQUE INDEX "Topics_name_key" ON public."Topics" USING btree (name, "deletedAt", "mapsTo")`

Reply:
I think this might be a better way to go. Make the unique index with "deletedAt" as part of the index. Then make a unique constraint using the unique index. It will keep all the original goals and still allow for reuse after delete.

Reply:
make sure to drop the old indexes too

Reply:
Original Indexes:
`	SELECT DISTINCT *
	FROM pg_indexes i
	join information_schema.columns isc
	on i.tablename = isc.table_name
	and isc.column_name = 'deletedAt'
	where tablename not like 'pg_%'
	and indexname not like '%_pkey'
	and indexdef ilike '% UNIQUE %'
	and tablename not like 'Monitoring%';`

Original constraints:
`SELECT
tc.table_name,
  'ALTER TABLE ' || tc.table_schema || '.' || tc.table_name || 
  ' ADD CONSTRAINT ' || tc.constraint_name || 
  ' UNIQUE (' || string_agg(kcu.column_name, ', ') || ');' as create_statement
FROM 
  information_schema.table_constraints AS tc
JOIN 
  information_schema.key_column_usage AS kcu
  ON tc.constraint_name = kcu.constraint_name
join information_schema.columns isc
	on tc.table_name = isc.table_name
	and isc.column_name = 'deletedAt' 
WHERE 
  tc.constraint_type = 'UNIQUE'
GROUP BY 
  tc.table_schema, tc.table_name, tc.constraint_name;`

Reply:
I think which approach is better depends on whether we expect to actually see the deleted records in any queries. If they're always excluded, then a smaller index with only the non-excluded records is usually better.

Reply:
Due to ON CONFLICT with upserts not working with partial indexes, I am switching to including `deletedAt` in the indexes regardless of other considerations

Reply:
Actually, adding `deletedAt` doesn't work either because it means the conflict column specification doesn't match the new constraint. I think we may need to do a legit undelete

Reply:
We could have gone back to using just partial indexes because @thewatermethod removed the upsert, but I think enough folks are waiting on this to where we just want to merge